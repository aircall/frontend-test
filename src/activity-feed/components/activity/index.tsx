import React from "react";
import { useRecoilValue } from "recoil";
import { activityByIdState } from "../../selectors";

export type Props = {
  id: number;
};

export function Activity({ id }: Props) {
  const activity = useRecoilValue(activityByIdState(id));

  return <section>{activity?.from}</section>;
}
