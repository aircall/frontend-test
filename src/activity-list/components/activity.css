.activity {
    display: flex;
    flex-direction: row;
    box-shadow: 0 3px 1px -2px rgba(0,0,0,.2), 0 2px 2px 0 rgba(0,0,0,.14), 0 1px 5px 0 rgba(0,0,0,.1);
    padding: 10px;
    margin-bottom: 20px;
    align-items: center;
}
.activity div {
    display: flex;
    flex-direction: column;
    margin-right: 20px;
}
.activity_details {
    flex: 1;
}
.activity_details span {
    line-height: 20px;
}
.activity_more-action {
    font-size: 20px;
    cursor: pointer;
}