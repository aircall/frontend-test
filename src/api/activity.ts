import { apiCall } from ".";
import { Activity } from "../models/Activity";
import * as t from "io-ts";

export const indexActivities = () => {
  return apiCall({
    method: "GET",
    path: "/activities",
    validator: t.array(Activity)
  });
};

export const showActivity = (id: number) => {
  return apiCall({
    method: "GET",
    path: `/activities/${id}`,
    validator: Activity
  });
};

export const updateActivity = (
  id: number,
  params: { is_archived: boolean }
) => {
  return apiCall({
    method: "POST",
    path: `/activities/${id}`,
    validator: Activity,
    params
  });
};

export const resetActivities = () => {
  return apiCall({
    method: "GET",
    path: "/reset"
  });
};
