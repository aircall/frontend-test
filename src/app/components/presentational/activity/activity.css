.activity {
  position: relative;
  display: block;
  padding: 20px;
  margin-bottom: 15px;
  transition: background-color 0.3s ease-in;
  border: 1px solid #EEEEED;
  background-color: #FFFFFF;
  border-radius: 14px;
  text-decoration: none;
  font-weight: 600;
  font-size: 1.4rem;
  color: #000;
}

.activity:hover {
  background-color: #FDFDFC;
}

.activity:last-child {
  margin-bottom: 0;
}

.activity__time {
  position: absolute;
  top: 50%;
  right: -1px;
  padding: 5px 10px;
  transform: translateY(-50%);
  border: 1px solid #EEEEED;
  border-top-left-radius: 6px;
  border-bottom-left-radius: 6px;
  font-size: 1.2rem;
  color: #BABABA;
}

.activity__status {
  display: block;
  margin-top: 5px;
  font-size: 1rem;
  color: #B9B9B9;
}
