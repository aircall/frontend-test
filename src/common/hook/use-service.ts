import { useState, useEffect } from "react";

const noop = () => {};

/**
 * Calls an asynchronous service and keep track of loading state, data and errors
 * @param func Asynchronous function to call
 * @param callbacks
 * @param [callbacks.onSuccess] Function called in case of success
 * @param [callbacks.onError] Function called in case of error
 * @param deps Dependencies
 */
export function useService<T = any, E = Error>(
  func: () => Promise<T>,
  {
    onSuccess = noop,
    onError = noop,
  }: { onSuccess?: (data: T) => void; onError?: (error: E) => void } = {},
  deps: string[] = []
) {
  const [error, setError] = useState<E | null>(null);
  const [isLoading, setIsLoading] = useState(false);

  useEffect(() => {
    setError(null);
    setIsLoading(true);
    func()
      .then(onSuccess, (error: E) => {
        setError(error);
        onError(error);
      })
      .finally(() => setIsLoading(false));
  }, deps);

  return { error, isLoading };
}
