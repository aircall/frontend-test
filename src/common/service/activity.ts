import request from "axios";

export const API_URL = "https://aircall-job.herokuapp.com";

export enum CallType {
  MISSED = "missed",
  ANSWERED = "answered",
  VOICEMAIL = "voicemail",
}

export enum CallDirection {
  INBOUND = "inbound",
  OUTBOUND = "outbound",
}

export type Activity = {
  id: number;
  created_at: string;
  direction: CallDirection;
  from: string;
  to: string;
  via: string;
  duration: number;
  is_archived: boolean;
  call_type: CallType;
};

/**
 * Returns a list of activities
 */
export async function getActivityList(): Promise<Activity[]> {
  const response = await request.get<Activity[]>(`${API_URL}/activities`);
  return response.data;
}

/**
 * Return an activity by its ID
 * @param id ID of the activity we want to get
 */
export async function getActivityById(id: number): Promise<Activity> {
  const response = await request.get<Activity>(`${API_URL}/activities/${id}`);
  return response.data;
}

/**
 * Archive an activity
 * @param activity Activity to archive
 */
export async function archiveActivity(activity: Activity): Promise<void> {
  await request.post<void>(`${API_URL}/activities/${activity.id}`, {
    ...activity,
    is_archived: true,
  });
}
