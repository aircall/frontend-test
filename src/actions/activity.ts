import api from '../api'
import { IActivity } from '../shared/api-types'
import { KThunkAction } from './helpers'

export enum ActivityTypeKeys {
  ACTIVITIES_LIST_FETCHING = 'ACTIVITIES_LIST_FETCHING',
  ACTIVITIES_LIST_SUCCESS = 'ACTIVITIES_LIST_SUCCESS',
  ACTIVITIES_LIST_ERROR = 'ACTIVITIES_LIST_ERROR',

  ACTIVITY_FETCHING = 'ACTIVITY_FETCHING',
  ACTIVITY_SUCCESS = 'ACTIVITY_SUCCESS',
  ACTIVITY_ERROR = 'ACTIVITY_ERROR',

  UPDATE_ACTIVITY_FETCHING = 'UPDATE_ACTIVITY_FETCHING',
  UPDATE_ACTIVITY_SUCCESS = 'UPDATE_ACTIVITY_SUCCESS',
  UPDATE_ACTIVITY_ERROR = 'UPDATE_ACTIVITY_ERROR'
}

export type ActivityActions =
  | IActivitiesListFetchingAction
  | IActivitiesListSuccessAction
  | IActivitiesListErrorAction
  | IActivityFetchingAction
  | IActivitySuccessAction
  | IActivityErrorAction
  | IUpdateActivityFetchingAction
  | IUpdateActivitySuccessAction
  | IUpdateActivityErrorAction

// GET ACTIVITIES LIST
export interface IActivitiesListFetchingAction {
  type: ActivityTypeKeys.ACTIVITIES_LIST_FETCHING
}

export interface IActivitiesListSuccessAction {
  type: ActivityTypeKeys.ACTIVITIES_LIST_SUCCESS
  activities: IActivity[]
}

export interface IActivitiesListErrorAction {
  type: ActivityTypeKeys.ACTIVITIES_LIST_ERROR
  error: string
}

export function activitiesListFetching(): IActivitiesListFetchingAction {
  return {
    type: ActivityTypeKeys.ACTIVITIES_LIST_FETCHING
  }
}

export function activitiesListSuccess(activities: IActivity[]): IActivitiesListSuccessAction {
  return {
    type: ActivityTypeKeys.ACTIVITIES_LIST_SUCCESS,
    activities
  }
}

export function activitiesListError(error: string): IActivitiesListErrorAction {
  return {
    type: ActivityTypeKeys.ACTIVITIES_LIST_ERROR,
    error
  }
}

export function doGetActivitiesList(): KThunkAction {
  return async (dispatch) => {
    dispatch(activitiesListFetching())
    try {
      const activitiesResponse = await api.activity.getActivitiesList()
      dispatch(activitiesListSuccess(activitiesResponse.activities))
    } catch (err) {
      dispatch(activitiesListError(err))
    }
  }
}

// GET ACTIVITY BY ID
export interface IActivityFetchingAction {
  type: ActivityTypeKeys.ACTIVITY_FETCHING
}

export interface IActivitySuccessAction {
  type: ActivityTypeKeys.ACTIVITY_SUCCESS
  activity: IActivity
}

export interface IActivityErrorAction {
  type: ActivityTypeKeys.ACTIVITY_ERROR
  error: string
}

export function activityFetching(): IActivityFetchingAction {
  return {
    type: ActivityTypeKeys.ACTIVITY_FETCHING
  }
}

export function activitySuccess(activity: IActivity): IActivitySuccessAction {
  return {
    type: ActivityTypeKeys.ACTIVITY_SUCCESS,
    activity
  }
}

export function activityError(error: string): IActivityErrorAction {
  return {
    type: ActivityTypeKeys.ACTIVITY_ERROR,
    error
  }
}

export function doGetActivity(id: string): KThunkAction {
  return async (dispatch) => {
    dispatch(activityFetching())
    try {
      const activityResponse = await api.activity.getActivityById(id)
      dispatch(activitySuccess(activityResponse.activity))
    } catch (err) {
      dispatch(activityError(err))
    }
  }
}

// UPDATE ACTIVITY
export interface IUpdateActivityFetchingAction {
  type: ActivityTypeKeys.UPDATE_ACTIVITY_FETCHING
}

export interface IUpdateActivitySuccessAction {
  type: ActivityTypeKeys.UPDATE_ACTIVITY_SUCCESS
  activity: IActivity
}

export interface IUpdateActivityErrorAction {
  type: ActivityTypeKeys.UPDATE_ACTIVITY_ERROR
  error: string
}

export function updateActivityFetching(): IUpdateActivityFetchingAction {
  return {
    type: ActivityTypeKeys.UPDATE_ACTIVITY_FETCHING
  }
}

export function updateActivitySuccess(activity: IActivity): IUpdateActivitySuccessAction {
  return {
    type: ActivityTypeKeys.UPDATE_ACTIVITY_SUCCESS,
    activity
  }
}

export function updateActivityError(error: string): IUpdateActivityErrorAction {
  return {
    type: ActivityTypeKeys.UPDATE_ACTIVITY_ERROR,
    error
  }
}

export function doUpdateActivity(id: string, is_archived: boolean): KThunkAction {
  return async (dispatch) => {
    dispatch(updateActivityFetching())
    try {
      const activityResponse = await api.activity.updateActivity(id, is_archived)
      dispatch(updateActivitySuccess(activityResponse.activity))
    } catch (err) {
      dispatch(updateActivityError(err))
    }
  }
}
