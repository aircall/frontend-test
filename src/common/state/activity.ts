import { atom } from "recoil";

import { Activity } from "../service/activity";

export const listState = atom<Activity[]>({
  key: "activityList",
  default: [],
});

export const detailState = atom<Activity | null>({
  key: "activityDetail",
  default: null,
});
