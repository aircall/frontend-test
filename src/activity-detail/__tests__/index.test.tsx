import React from "react";
import { useRecoilState } from "recoil";

import { render } from "testing-utils";

import { ActivityDetail, Props } from "..";
import { useService } from "../../common/hook/use-service";
import { getActivityById } from "../../common/service/activity";

jest.mock("recoil");
jest.mock("../../common/hook/use-service");
jest.mock("../../common/service/activity");

describe("activity-detail", () => {
  describe("ActivityDetail", () => {
    const useServiceMock = useService as jest.Mock;
    const useRecoilStateMock = useRecoilState as jest.Mock;
    const getActivityByIdMock = getActivityById as jest.Mock;

    const fakeRecoilState = {
      id: 7834,
      created_at: "2018-04-19T09:38:41.000Z",
      direction: "outbound",
      from: "Pierre-Baptiste BÃ©chu",
      to: "06 46 62 12 33",
      via: "NYC Office",
      duration: "120",
      is_archived: false,
      call_type: "answered",
    };

    describe("#render", () => {
      it("should render correctly the activity", () => {
        useRecoilStateMock.mockReturnValueOnce([fakeRecoilState, () => {}]);
        useServiceMock.mockReturnValueOnce({ isLoading: false, error: null });

        const props = {
          match: { params: { id: "42" } },
        } as Props;
        const { container } = render(<ActivityDetail {...props} />);

        expect(container).toMatchSnapshot();
      });

      it("should render correctly while loading", () => {
        useRecoilStateMock.mockReturnValueOnce([null, () => {}]);
        useServiceMock.mockReturnValueOnce({
          isLoading: true,
          error: null,
        });

        const props = {
          match: { params: { id: "42" } },
        } as Props;
        const { container } = render(<ActivityDetail {...props} />);

        expect(container).toMatchSnapshot();
      });

      it("should render correctly when there's an error", () => {
        useRecoilStateMock.mockReturnValueOnce([null, () => {}]);
        useServiceMock.mockReturnValueOnce({
          isLoading: false,
          error: new Error("test_error"),
        });

        const props = {
          match: { params: { id: "42" } },
        } as Props;
        const { container } = render(<ActivityDetail {...props} />);

        expect(container).toMatchSnapshot();
      });
    });

    describe("#useService", () => {
      it("should call the service and set the activity detail in case of success", () => {
        const setActivityMock = jest.fn();
        useRecoilStateMock.mockReturnValueOnce([null, setActivityMock]);
        useServiceMock
          .mockImplementationOnce((serviceCaller, { onSuccess }) => {
            serviceCaller();
            onSuccess(fakeRecoilState);
            return { isLoading: false, error: null };
          })
          .mockReturnValueOnce({ isLoading: false, error: null });

        const props = {
          match: { params: { id: "42" } },
        } as Props;
        render(<ActivityDetail {...props} />);

        expect(getActivityByIdMock).toHaveBeenCalledWith(42);
        expect(setActivityMock).toHaveBeenCalledWith(fakeRecoilState);
      });
    });
  });
});
