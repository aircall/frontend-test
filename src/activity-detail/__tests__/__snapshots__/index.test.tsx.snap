// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`activity-detail ActivityDetail #render should render correctly the activity 1`] = `
<div>
  Pierre-Baptiste BÃ©chu
</div>
`;

exports[`activity-detail ActivityDetail #render should render correctly when there's an error 1`] = `
<div>
  test_error
</div>
`;

exports[`activity-detail ActivityDetail #render should render correctly while loading 1`] = `
<div>
  Loading...
</div>
`;
